### ISSUE TITLE
Can't set breakpoints on a line with an opening bracket (e.g. list or tuple)

### ISSUE THREAD

**Issue by NotWearingPants** [microsoft/debugpy#255]
## Environment data

-   VS Code version: 1.45.0
-   Extension version (available under the Extensions sidebar): 2020.5.78807
-   OS and version: Windows 10
-   Python version: Python 3.6.1 32bit
-   Type of virtual environment used: none
-   `"python.jediEnabled": false`
-   `"python.languageServer": "Microsoft"`

## Steps to reproduce:

1. Open a new python file in vscode
2. Enter the following code:
```python
while True:
	pass

a = 0
b = [
    0
]
```
3. Start debugging it with the default "python file" configuration
4. Try to put a breakpoint on the line `b = [`

## Expected behaviour

The breakpoint is placed on the line I pressed.

## Actual behaviour

The red breakpoint circle appears next to the line `b = [` for a split second, and then moved automatically to the line above.

If you try to place the breakpoint again, it will again move to the line above, and vscode will show it having 2 breakpoints on the same line.

## Notes

- The `while True` is irrelevant, it's there to have the program run so we can debug.

- This also happens if I change `b` to a tuple instead of a list:
```python
b = (
    0
)
```
- If we add a variable name before the bracket, then it works as expected, e.g.:
```python
b = a[
    0
]
```
or
```python
b = a(
    0
)
```

## Logs

<details>

<summary>Output for <code>Python</code> in the <code>Output</code> panel
</summary>

<p>

```
User belongs to experiment group 'AlwaysDisplayTestExplorer - experiment'
User belongs to experiment group 'ShowPlayIcon - start'
User belongs to experiment group 'DebugAdapterFactory - experiment'
User belongs to experiment group 'PtvsdWheels37 - experiment'
User belongs to experiment group 'UseTerminalToGetActivatedEnvVars - control'
User belongs to experiment group 'WebHostNotebook - control'
User belongs to experiment group 'LocalZMQKernel - control'
User belongs to experiment group 'NativeNotebook - control'
User belongs to experiment group 'CollectLSRequestTiming - experiment'
User belongs to experiment group 'CollectNodeLSRequestTiming - experiment'
User belongs to experiment group 'EnableIPyWidgets - experiment'
User belongs to experiment group 'RunByLine - control'
```

</p>
</details>

<br />

<details>

<summary>
Output from <code>Console</code> under the <code>Developer Tools</code> panel (toggle Developer Tools on under <code>Help</code>; turn on source maps to make any tracebacks be useful by running <code>Enable source map support for extension debugging</code>)
</summary>

<p>

```
```

</p>
</details>

**Comment by karthiknadig** [comment]
@NotWearingPants This is by design. The debugger automatically corrects the line depending on what python tells us is a valid line to break on. We try to go to previous line since it gives the user an opportunity to block before the intended line. Say your code:
```
Line 1: a = 0
Line 2: b = [
Line 3: 0
Line 4: ]
```
According to python 2.7 only line 1 and 3 are breakable (first column is line:
```
py -2.7 C:\GIT\issues\demo_feb\skiplines.py
  1           0 LOAD_CONST               0 (0)
              3 STORE_NAME               0 (a)

  3           6 LOAD_CONST               0 (0)
              9 BUILD_LIST               1
             12 STORE_NAME               1 (b)
             15 LOAD_CONST               1 (None)
             18 RETURN_VALUE
```
For python 3.6 which is what you are using, it is Line 1, 3 (similar to 2.7):
```
py -3.6 C:\GIT\issues\demo_feb\skiplines.py
  1           0 LOAD_CONST               0 (0)
              2 STORE_NAME               0 (a)

  3           4 LOAD_CONST               0 (0)
              6 BUILD_LIST               1
              8 STORE_NAME               1 (b)
             10 LOAD_CONST               1 (None)
             12 RETURN_VALUE
```
Python 3.8 handles this differently, allows breaking on line 2:
```
py -3.8 C:\GIT\issues\demo_feb\skiplines.py
  1           0 LOAD_CONST               0 (0)
              2 STORE_NAME               0 (a)

  3           4 LOAD_CONST               0 (0)

  2           6 BUILD_LIST               1
              8 STORE_NAME               1 (b)
             10 LOAD_CONST               1 (None)
             12 RETURN_VALUE
```
On 3.8 you should be able to set breakpoint on line 2 and it will not adjust it.
![image](https://user-images.githubusercontent.com/3840081/82473210-15c4d780-9a7e-11ea-8253-4161813edd94.png)

**Comment by NotWearingPants** [comment]
Thanks for replying, and thanks for the analysis, this bug is pretty weird admittedly.
Do you know where this change in Python 3.8 was noted? or the issue number?

But regardless of the python bug that was fixed, there's still an issue with your debug adapter - auto-correcting the line of the breakpoint can cause it to have 2 breakpoints on the same line.

Though I'm not sure if this is a bug with the debug adapter or with VSCode showing it wrong.

Thoughts?

**Comment by fabioz** [comment]
I believe that having 2 breakpoints on the same line (after one was moved from a line which wouldn't be hit to a valid line) isn't really a bug and it seems to be the correct thing to do (after the debug session the lines are adjusted to the initial lines) -- also, this behavior is dictated by the UI, not the debugger implementation.

Related to Python 3.8, this was a change in the bytecode generated (in this case, after 0 is loaded then it backtracks to the previous line before doing the assign -- it isn't mentioned in the docs: https://docs.python.org/3/whatsnew/3.8.html, but it definitely was a new feature added in 3.8 -- I had created an issue myself when I noted that some time ago. See: https://bugs.python.org/issue38508).

p.s.: I'm closing this issue as there's nothing to be done in the debugger in this case (everything seems to be working as designed).
